// arduino-ide-extension/lib/node/arduino-daemon.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = require("path");
var child_process_1 = require("child_process");
var inversify_1 = require("inversify");
var logger_1 = require("@theia/core/lib/common/logger");
var promise_util_1 = require("@theia/core/lib/common/promise-util");
var environment_1 = require("@theia/application-package/lib/environment");
var daemon_log_1 = require("./daemon-log");
var tool_output_service_1 = require("../common/protocol/tool-output-service");
var arduino_cli_contribution_1 = require("./arduino-cli-contribution");
var arduino_cli_1 = require("./arduino-cli");
var ArduinoDaemon = /** @class */ (function () {
    function ArduinoDaemon() {
        this.isReady = new promise_util_1.Deferred();
    }
    ArduinoDaemon.prototype.onStart = function () {
        return __awaiter(this, void 0, void 0, function () {
            var executable, version, daemon, options, command, cp, error_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 5, , 6]);
                        if (!!this.cliContribution.debugCli) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.cli.getExecPath()];
                    case 1:
                        executable = _a.sent();
                        return [4 /*yield*/, this.cli.getVersion()];
                    case 2:
                        version = _a.sent();
                        this.logger.info(">>> Starting " + version.toLocaleLowerCase() + " daemon from " + executable + "...");
                        daemon = child_process_1.exec("\"" + executable + "\" daemon -v --log-level info --format json --log-format json", { encoding: 'utf8', maxBuffer: 1024 * 1024 }, function (err, stdout, stderr) {
                            if (err || stderr) {
                                console.log(err || new Error(stderr));
                                return;
                            }
                            console.log(stdout);
                        });
                        options = {
                            env: environment_1.environment.electron.runAsNodeEnv(),
                            detached: true,
                            stdio: 'ignore'
                        };
                        command = process.execPath;
                        cp = child_process_1.spawn(command, [path_1.join(__dirname, 'daemon-watcher.js'), String(process.pid), String(daemon.pid)], options);
                        cp.unref();
                        if (daemon.stdout) {
                            daemon.stdout.on('data', function (data) {
                                _this.toolOutputService.publishNewOutput('daemon', daemon_log_1.DaemonLog.toPrettyString(data.toString()));
                                daemon_log_1.DaemonLog.log(_this.logger, data.toString());
                            });
                        }
                        if (daemon.stderr) {
                            daemon.stderr.on('data', function (data) {
                                _this.toolOutputService.publishNewOutput('daemon error', daemon_log_1.DaemonLog.toPrettyString(data.toString()));
                                daemon_log_1.DaemonLog.log(_this.logger, data.toString());
                            });
                        }
                        if (daemon.stderr) {
                            daemon.on('exit', function (code, signal) { return daemon_log_1.DaemonLog.log(_this.logger, "Daemon exited with code: " + code + ". Signal was: " + signal + "."); });
                        }
                        _a.label = 3;
                    case 3: return [4 /*yield*/, new Promise(function (resolve) { return setTimeout(resolve, 2000); })];
                    case 4:
                        _a.sent();
                        this.isReady.resolve();
                        if (!this.cliContribution.debugCli) {
                            this.logger.info("<<< The 'arduino-cli' daemon is up and running.");
                        }
                        else {
                            this.logger.info("Assuming the 'arduino-cli' already runs in debug mode.");
                        }
                        return [3 /*break*/, 6];
                    case 5:
                        error_1 = _a.sent();
                        this.isReady.reject(error_1 || new Error('failed to start arduino-cli'));
                        return [3 /*break*/, 6];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    __decorate([
        inversify_1.inject(logger_1.ILogger),
        inversify_1.named('daemon'),
        __metadata("design:type", Object)
    ], ArduinoDaemon.prototype, "logger", void 0);
    __decorate([
        inversify_1.inject(arduino_cli_1.ArduinoCli),
        __metadata("design:type", arduino_cli_1.ArduinoCli)
    ], ArduinoDaemon.prototype, "cli", void 0);
    __decorate([
        inversify_1.inject(arduino_cli_contribution_1.ArduinoCliContribution),
        __metadata("design:type", arduino_cli_contribution_1.ArduinoCliContribution)
    ], ArduinoDaemon.prototype, "cliContribution", void 0);
    __decorate([
        inversify_1.inject(tool_output_service_1.ToolOutputServiceServer),
        __metadata("design:type", Object)
    ], ArduinoDaemon.prototype, "toolOutputService", void 0);
    ArduinoDaemon = __decorate([
        inversify_1.injectable()
    ], ArduinoDaemon);
    return ArduinoDaemon;
}());
exports.ArduinoDaemon = ArduinoDaemon;
