// arduino-ide-extension/lib/node/fs-extra.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var util_1 = require("util");
exports.existsSync = fs.existsSync;
exports.lstatSync = fs.lstatSync;
exports.readdirSync = fs.readdirSync;
exports.statSync = fs.statSync;
exports.writeFileSync = fs.writeFileSync;
exports.exists = util_1.promisify(fs.exists);
exports.lstat = util_1.promisify(fs.lstat);
exports.readdir = util_1.promisify(fs.readdir);
exports.stat = util_1.promisify(fs.stat);
exports.writeFile = util_1.promisify(fs.writeFile);
function mkdirp(path, timeout) {
    if (timeout === void 0) { timeout = 3000; }
    return new Promise(function (resolve, reject) {
        var timeoutHandle;
        if (timeout > 0) {
            timeoutHandle = setTimeout(function () {
                reject(new Error("Timeout of " + timeout + " ms exceeded while trying to create the directory \"" + path + "\""));
            }, timeout);
        }
        fs.mkdir(path, { recursive: true }, function (err) {
            clearTimeout(timeoutHandle);
            if (err)
                reject(err);
            else
                resolve();
        });
    });
}
exports.mkdirp = mkdirp;
function mkdirpSync(path) {
    fs.mkdirSync(path, { recursive: true });
}
exports.mkdirpSync = mkdirpSync;
