// arduino-ide-extension/lib/node/tool-output-service-impl.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var ToolOutputServiceServerImpl = /** @class */ (function () {
    function ToolOutputServiceServerImpl() {
        this.clients = [];
    }
    ToolOutputServiceServerImpl.prototype.publishNewOutput = function (tool, chunk) {
        if (!chunk) {
            return;
        }
        this.clients.forEach(function (c) { return c.onNewOutput(tool, chunk); });
    };
    ToolOutputServiceServerImpl.prototype.setClient = function (client) {
        if (!client) {
            return;
        }
        this.clients.push(client);
    };
    ToolOutputServiceServerImpl.prototype.disposeClient = function (client) {
    };
    ToolOutputServiceServerImpl.prototype.dispose = function () {
        this.clients = [];
    };
    ToolOutputServiceServerImpl = __decorate([
        inversify_1.injectable()
    ], ToolOutputServiceServerImpl);
    return ToolOutputServiceServerImpl;
}());
exports.ToolOutputServiceServerImpl = ToolOutputServiceServerImpl;
