// arduino-ide-extension/lib/browser/customization/arduino-editor-contribution.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var editor_contribution_1 = require("@theia/editor/lib/browser/editor-contribution");
var browser_1 = require("@theia/core/lib/browser");
var ArduinoEditorContribution = /** @class */ (function (_super) {
    __extends(ArduinoEditorContribution, _super);
    function ArduinoEditorContribution() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ArduinoEditorContribution.prototype.updateLanguageStatus = function (editor) {
    };
    ArduinoEditorContribution.prototype.setCursorPositionStatus = function (editor) {
        if (!editor) {
            this.statusBar.removeElement('editor-status-cursor-position');
            return;
        }
        var cursor = editor.cursor;
        this.statusBar.setElement('editor-status-cursor-position', {
            text: "" + (cursor.line + 1),
            alignment: browser_1.StatusBarAlignment.LEFT,
            priority: 100
        });
    };
    return ArduinoEditorContribution;
}(editor_contribution_1.EditorContribution));
exports.ArduinoEditorContribution = ArduinoEditorContribution;
