// arduino-ide-extension/lib/browser/library/library-list-widget.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var library_service_1 = require("../../common/protocol/library-service");
var list_widget_1 = require("../components/component-list/list-widget");
var list_item_renderer_1 = require("../components/component-list/list-item-renderer");
var LibraryListWidget = /** @class */ (function (_super) {
    __extends(LibraryListWidget, _super);
    function LibraryListWidget(service, itemRenderer) {
        var _this = _super.call(this, {
            id: LibraryListWidget_1.WIDGET_ID,
            label: LibraryListWidget_1.WIDGET_LABEL,
            iconClass: 'library-tab-icon',
            searchable: service,
            installable: service,
            itemLabel: function (item) { return item.name; },
            itemRenderer: itemRenderer
        }) || this;
        _this.service = service;
        _this.itemRenderer = itemRenderer;
        return _this;
    }
    LibraryListWidget_1 = LibraryListWidget;
    var LibraryListWidget_1;
    LibraryListWidget.WIDGET_ID = 'library-list-widget';
    LibraryListWidget.WIDGET_LABEL = 'Library Manager';
    LibraryListWidget = LibraryListWidget_1 = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(library_service_1.LibraryService)),
        __param(1, inversify_1.inject(list_item_renderer_1.ListItemRenderer)),
        __metadata("design:paramtypes", [Object, list_item_renderer_1.ListItemRenderer])
    ], LibraryListWidget);
    return LibraryListWidget;
}(list_widget_1.ListWidget));
exports.LibraryListWidget = LibraryListWidget;
